src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
+                {
+                    // We need to Parse this as a String instead. There are cases
+                    // when you tlbimp enums that can have values of the form "3D".
                 }
-                catch (Exception ex)
+                catch when (!throwOnFailure)
                 {
-                    if (parseResult.canThrow)
-                        throw;
-                    else
-                    {
-                        parseResult.SetFailure(ex);
-                        return false;
-                    }
+                    enumResult = default;
+                    return false;
                 }
             }
 
-            // Find the field. Let's assume that these are always static classes 
-            // because the class is an enum.
-            TypeValuesAndNames entry = GetCachedValuesAndNames(rtType, true);
+            if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulong ulongResult))
+            {
+                try
+                {
+                    enumResult = (TResult)ToObject(enumType, ulongResult);
+                    return true;
+                }
+                catch when (!throwOnFailure) { }
+            }
+
+            enumResult = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool StartsNumber(char c) => char.IsInRange(c, '0', '9') || c == '-' || c == '+';
+
+        private static bool TryParseByName(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out ulong ulongResult)
+        {
+            // Find the field. Let's assume that these are always static classes because the class is an enum.
+            TypeValuesAndNames entry = GetCachedValuesAndNames(enumType, getNames: true);
             string[] enumNames = entry.Names;
             ulong[] enumValues = entry.Values;
 
-            StringComparison comparison = ignoreCase ?
-                StringComparison.OrdinalIgnoreCase :
-                StringComparison.Ordinal;
-
-            int valueIndex = firstNonWhitespaceIndex;
-            while (valueIndex <= value.Length) // '=' is to handle invalid case of an ending comma
+            bool parsed = true;
+            ulong result = 0;
+            while (value.Length > 0)
             {
-                // Find the next separator, if there is one, otherwise the end of the string.
-                int endIndex = value.IndexOf(enumSeparatorChar, valueIndex);
+                // Find the next separator.
+                ReadOnlySpan<char> subvalue;
+                int endIndex = value.IndexOf(enumSeparatorChar);
                 if (endIndex == -1)
                 {
-                    endIndex = value.Length;
+                    // No next separator; use the remainder as the next value.
+                    subvalue = value.Trim();
```

Comment:
Do we need to trim here or is `TrimStart` sufficient?

Reply:
We need Trim.  You could have values like " hi   , there    , world", and we need to trim off the spaces e.g. before and after "there".

Reply:
Right, that makes sense. So, why do we have TrimStart only in the other cases (which probably results in an extra iteration within TryParseByName)?

For example: https://github.com/dotnet/coreclr/pull/21214/files#diff-b640a1b12cbce0219cf8f94d46318186R326

Reply:
Because in those cases we only care that we don't begin with whitespace, so that when we get to a line like https://github.com/dotnet/coreclr/pull/21214/files#diff-b640a1b12cbce0219cf8f94d46318186R486, we can just check str[0] to determine whether it's a number or not, and then the number parsing doesn't pass in AllowLeadingWhite because it's known to have already been trimmed.

Reply:
But if that case isn't hit, we pass the trailing whitespace `value` to TryParseByName where we are iterating over the "substrings" using `while (value.Length > 0)`. After the last `EnumSeparatorChar` separated value, we still have `value.Length > 0`, no?

Reply:
> After the last EnumSeparatorChar separated value, we still have value.Length > 0, no?

How is that related and why is that a problem?

Reply:
Ok :) Thanks